[BEAM-3042] Tracking of time spent reading side inputs, and bytes read in Dataflow.

pabloem opened the pull request:
This pull request adds changes the PrefetchingSourceReader used to fetch side input data. 

Specifically, this change helps it track how long was spent blocked waiting to fetch side inputs; and for the Dataflow runner (which uses NativeAvroSource), it also helps track how many bytes were read.

pabloem commented:
r: @charlesccychen 
Hi Charles! I've tried to add enough documentation to my changes. If you need, I can detail to you what the purpose is for them.

charlesccychen commented:
Hi Pablo, thank you for this change.  If possible, I would like to avoid adding code that depends on internal Dataflow worker details.  I suggest that for the reader whose byte count you would like to observe, you inherit from ObservableMixin (see apache_beam/coders/observable.py) and register an observer for your particular use case.

pabloem commented:
That's a great tip, Charles. I've incorporated it into the change.

pabloem commented:
retest this please

pabloem commented:
Python tests are passing.

pabloem commented:
I've placed my changes behind an experiment flag, to test the change before full release.

pabloem commented:
retest this please

pabloem commented:
retest this please

pabloem commented:
I've pushed my new changes, along with test, and cythonization. This should be ready for review.

pabloem commented:
I've made sure that the performance impact is zero when the experiment is not activated (by not adding an observer if that's the case). Let me know what you think!
-P.

charlesccychen commented on sdks/python/apache_beam/runners/worker/opcounters.pxd:
```diff
@@ -21,6 +21,22 @@ cimport libc.stdint
 from apache_beam.utils.counters cimport Counter
 
 
+cdef class TransformIoCounter(object):
```
<!--new_thread; commit:4255543cdc7cfcadd29633f321599d2f591021d8; resolved:0-->
Let's capitalize IO in the name, i.e. `TransformIOCounter` (for consistency with, for example, `cStringIO` in the Python standard library).

pabloem replied:
<!--thread_id:cc_158386090_t; commit:4255543cdc7cfcadd29633f321599d2f591021d8; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Let's capitalize IO in the name, i.e. `TransformIOCounter` (for consistency with, for example, `cStringIO` in the Python standard library).

<!--section:body-->
Done.

charlesccychen commented on sdks/python/apache_beam/runners/worker/sideinputs.py:
```diff
@@ -130,7 +163,14 @@ def __iter__(self):
+            # be able to get from it instantly). This allows us to limit how
+            # much we check the current step.
+            self.read_counter.check_step()
+          with self.read_counter:
```
<!--new_thread; commit:4255543cdc7cfcadd29633f321599d2f591021d8; resolved:0-->
It doesn't look like to me that the changes here are guarded by the experiment flag?  Perhaps we can run a quick profile to verify these new changes don't change performance (particularly here with the "with self.read_counter").

pabloem replied:
<!--thread_id:cc_158386091_t; commit:4255543cdc7cfcadd29633f321599d2f591021d8; resolved:0-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> It doesn't look like to me that the changes here are guarded by the experiment flag?  Perhaps we can run a quick profile to verify these new changes don't change performance (particularly here with the "with self.read_counter").

<!--section:body-->
Will do.

pabloem replied:
<!--thread_id:cc_158386091_t; commit:4255543cdc7cfcadd29633f321599d2f591021d8; resolved:0-->
<!--section:context-quote-->
> **pabloem** wrote:
> Will do.

<!--section:body-->
I have run two of the Python Benchmark tests - I ran each job 10 times with each configuration (baseline, change behind experiment, change activated). The results are:

sideinputs_iterable_benchmark:
Baseline: Mean: 401s, Stdev: 13s.
With Deactivated Change: Mean: 533s, Stdev 18s.
With Change: Mean: 660s, Stdev: 60s

wordcount_many_files_1gb (which uses sideinputs in the write step)
Baseline: Mean: 560s, Stdev: 40s
With Deactivated Change: 567s, Stdev 21s.
With Change: 548s, Stdev: 17s.

I would conclude that the change does definitely affect sideinput performance, but this does not represent a significant change for the wordcount pipeline. I can try similar benchmarks with another integration test that you may recommend.
-P.

pabloem replied:
<!--thread_id:cc_158386091_t; commit:4255543cdc7cfcadd29633f321599d2f591021d8; resolved:1-->
<!--section:context-quote-->
> **pabloem** wrote:
> I have run two of the Python Benchmark tests - I ran each job 10 times with each configuration (baseline, change behind experiment, change activated). The results are:
> 
> sideinputs_iterable_benchmark:
> Baseline: Mean: 401s, Stdev: 13s.
> With Deactivated Change: Mean: 533s, Stdev 18s.
> With Change: Mean: 660s, Stdev: 60s
> 
> wordcount_many_files_1gb (which uses sideinputs in the write step)
> Baseline: Mean: 560s, Stdev: 40s
> With Deactivated Change: 567s, Stdev 21s.
> With Change: 548s, Stdev: 17s.
> 
> I would conclude that the change does definitely affect sideinput performance, but this does not represent a significant change for the wordcount pipeline. I can try similar benchmarks with another integration test that you may recommend.
> -P.

<!--section:body-->
I've removed the change that increased the main fetching path for side inputs. The performance impact with experiment deactivated will be zero now.

charlesccychen commented on sdks/python/apache_beam/runners/worker/opcounters.py:
```diff
@@ -24,11 +24,99 @@
+      state_sampler: A statesampler.StateSampler to transition into read states.
+      declaring_step: The step that directly receives the side input initially.
+      input_index: The index of the side input in the list of inputs of the
+        declaring step.
```
<!--new_thread; commit:1e64e1f8b880a9fe4450c84cf1c7c5694eb3567e; resolved:0-->
Is this a stable thing to refer to, i.e. is the order of the side inputs guaranteed?  I recall that side inputs are keyed by a string label / key in `si_tags_and_types` (see https://github.com/apache/beam/blob/1882939c0c240d8d1fff654f3f46be59b87ba1d2/sdks/python/apache_beam/runners/dataflow/dataflow_runner.py#L608).

pabloem replied:
<!--thread_id:cc_158386095_t; commit:1e64e1f8b880a9fe4450c84cf1c7c5694eb3567e; resolved:0-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Is this a stable thing to refer to, i.e. is the order of the side inputs guaranteed?  I recall that side inputs are keyed by a string label / key in `si_tags_and_types` (see https://github.com/apache/beam/blob/1882939c0c240d8d1fff654f3f46be59b87ba1d2/sdks/python/apache_beam/runners/dataflow/dataflow_runner.py#L608).

<!--section:body-->
Yes. The order of the side inputs is guaranteed. 

In the design phase of this project, we decided to use the (declaring_step, input index) pair to identify a side input because it's a concept that fits well within DFE. PCollections only have a Tag associated in Dataflow (not in Flume). We can get on GVC and have a chat about the design considerations if you'd like.

pabloem replied:
<!--thread_id:cc_158386095_t; commit:1e64e1f8b880a9fe4450c84cf1c7c5694eb3567e; resolved:1-->
<!--section:context-quote-->
> **pabloem** wrote:
> Yes. The order of the side inputs is guaranteed. 
> 
> In the design phase of this project, we decided to use the (declaring_step, input index) pair to identify a side input because it's a concept that fits well within DFE. PCollections only have a Tag associated in Dataflow (not in Flume). We can get on GVC and have a chat about the design considerations if you'd like.

<!--section:body-->
Done.

charlesccychen commented on sdks/python/apache_beam/runners/worker/opcounters.py:
```diff
@@ -24,11 +24,99 @@
+
+  This class is designed to track consumption of side inputs across fused steps.
+  We represent a side input as a declaring step, and an input index, which are
+  represented as an io_target.
```
<!--new_thread; commit:1e64e1f8b880a9fe4450c84cf1c7c5694eb3567e; resolved:0-->
Do we define what we mean by an `io_target` elsewhere?

Please add extra newline after paragraph.

pabloem replied:
Done.

charlesccychen commented on sdks/python/apache_beam/runners/worker/opcounters.py:
```diff
@@ -24,11 +24,99 @@
+    Args:
+      counter_factory: A counters.CounterFactory to create byte counters.
+      state_sampler: A statesampler.StateSampler to transition into read states.
+      declaring_step: The step that directly receives the side input initially.
```
<!--new_thread; commit:1e64e1f8b880a9fe4450c84cf1c7c5694eb3567e; resolved:0-->
What is the type of this?

pabloem replied:
<!--thread_id:cc_158386101_t; commit:1e64e1f8b880a9fe4450c84cf1c7c5694eb3567e; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> What is the type of this?

<!--section:body-->
Done.

charlesccychen commented on sdks/python/apache_beam/runners/worker/opcounters.py:
```diff
@@ -24,11 +24,99 @@
+    self._state_sampler = state_sampler
+    self.declaring_step = declaring_step
+    self.input_index = input_index
+    self.check_step()
```
<!--new_thread; commit:1e64e1f8b880a9fe4450c84cf1c7c5694eb3567e; resolved:0-->
Why do we do this here?--when do we construct this class, and how do we know that is where the data is consumed (per the details of `check_step()` you have below)?

pabloem replied:
<!--thread_id:cc_158386102_t; commit:1e64e1f8b880a9fe4450c84cf1c7c5694eb3567e; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Why do we do this here?--when do we construct this class, and how do we know that is where the data is consumed (per the details of `check_step()` you have below)?

<!--section:body-->
Done.

charlesccychen commented on sdks/python/apache_beam/runners/worker/opcounters.py:
```diff
@@ -24,11 +24,99 @@
+    Due to the fusion optimization, user code may choose to emit the data
+    structure that holds side inputs (Iterable, Dict, or others). This call
+    updates the current step, to attribute the data consumption to the step
+    that is responsible for it.
```
<!--new_thread; commit:1e64e1f8b880a9fe4450c84cf1c7c5694eb3567e; resolved:0-->
"responsible for the actual consumption." ?

pabloem replied:
<!--thread_id:cc_158386103_t; commit:1e64e1f8b880a9fe4450c84cf1c7c5694eb3567e; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> "responsible for the actual consumption." ?

<!--section:body-->
Done.

charlesccychen commented on sdks/python/apache_beam/runners/worker/opcounters.py:
```diff
@@ -24,11 +24,99 @@
+  def __enter__(self):
+    pass
+
+  def add_bytes_read(self, n):
```
<!--new_thread; commit:1e64e1f8b880a9fe4450c84cf1c7c5694eb3567e; resolved:0-->
Let's change `n` to `count`.

pabloem replied:
<!--thread_id:cc_158386104_t; commit:1e64e1f8b880a9fe4450c84cf1c7c5694eb3567e; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Let's change `n` to `count`.

<!--section:body-->
Done.

charlesccychen commented on sdks/python/apache_beam/runners/worker/opcounters.py:
```diff
@@ -24,11 +24,99 @@
+  def check_step(self):
+    pass
+
+  def __exit__(self, exc_type, exc_value, traceback):
```
<!--new_thread; commit:1e64e1f8b880a9fe4450c84cf1c7c5694eb3567e; resolved:0-->
Can you document in the class docstring how this is meant to be used?  Are we supposed to use this class as a context manager, and if so, when do we enter and exit this context?

pabloem replied:
<!--thread_id:cc_158386105_t; commit:1e64e1f8b880a9fe4450c84cf1c7c5694eb3567e; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Can you document in the class docstring how this is meant to be used?  Are we supposed to use this class as a context manager, and if so, when do we enter and exit this context?

<!--section:body-->
Done.

charlesccychen commented on sdks/python/apache_beam/runners/worker/opcounters.py:
```diff
@@ -24,11 +24,99 @@
+  Examples of IO can be side inputs, shuffle, or streaming state.
+  """
+
+  def check_step(self):
```
<!--new_thread; commit:1e64e1f8b880a9fe4450c84cf1c7c5694eb3567e; resolved:0-->
Can you please document this method?

pabloem replied:
<!--thread_id:cc_158386106_t; commit:1e64e1f8b880a9fe4450c84cf1c7c5694eb3567e; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Can you please document this method?

<!--section:body-->
Done.

charlesccychen commented on sdks/python/apache_beam/runners/worker/opcounters.py:
```diff
@@ -24,11 +24,99 @@
+            io_target=counters.side_input_id(operation_name, self.input_index)),
+        Counter.SUM)
+
+  def add_bytes_read(self, n):
```
<!--new_thread; commit:1e64e1f8b880a9fe4450c84cf1c7c5694eb3567e; resolved:0-->
Let's change `n` to `count`.

pabloem replied:
<!--thread_id:cc_158386107_t; commit:1e64e1f8b880a9fe4450c84cf1c7c5694eb3567e; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Let's change `n` to `count`.

<!--section:body-->
Done.

charlesccychen commented on sdks/python/apache_beam/runners/worker/operations.py:
```diff
@@ -299,7 +299,14 @@ def _read_side_inputs(self, tags_and_types):
+          self.state_sampler,
+          declaring_step=self.operation_name,
+          # Inputs are 1-indexed, so we add 1 to i in the side input id
+          input_index=i + 1)
```
<!--new_thread; commit:0bfc12f8986238da5b5014cfc5eea4d32e6638f7; resolved:0-->
Please see comment in other file (opcounters.py) about stability of the order here (since we use side_tag) as the key for side inputs.

pabloem replied:
<!--thread_id:cc_158386110_t; commit:0bfc12f8986238da5b5014cfc5eea4d32e6638f7; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Please see comment in other file (opcounters.py) about stability of the order here (since we use side_tag) as the key for side inputs.

<!--section:body-->
Marking as done to keep discussion in opcounters.py.

charlesccychen commented on sdks/python/apache_beam/runners/worker/operations.py:
```diff
@@ -299,7 +299,14 @@ def _read_side_inputs(self, tags_and_types):
           raise NotImplementedError('Unknown side input type: %r' % si)
         sources.append(si.source)
-      iterator_fn = sideinputs.get_iterator_fn_for_sources(sources)
+      si_counter = opcounters.SideInputReadCounter(
```
<!--new_thread; commit:0bfc12f8986238da5b5014cfc5eea4d32e6638f7; resolved:0-->
It doesn't look like to me that the changes here are guarded by the experiment flag?  Perhaps we can run a quick profile to verify these new changes don't change performance (particularly with the "with self.read_counter" in side_inputs.py).

pabloem replied:
<!--thread_id:cc_158386111_t; commit:0bfc12f8986238da5b5014cfc5eea4d32e6638f7; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> It doesn't look like to me that the changes here are guarded by the experiment flag?  Perhaps we can run a quick profile to verify these new changes don't change performance (particularly with the "with self.read_counter" in side_inputs.py).

<!--section:body-->
You're right. I've added them now - and I'll run the profile tests in a bit. Marking as Done to track in comment within sideinputs.py.

charlesccychen commented on sdks/python/apache_beam/runners/worker/statesampler.pyx:
```diff
@@ -190,6 +195,13 @@ cdef class StateSampler(object):
+
+    This operation is not thread safe, and should only be used to check, not to
+    update information in the current state."""
+    return self.scoped_states_by_index[self.current_state_index]
```
<!--new_thread; commit:3e784cbd5e110c68b56b1b4349510d712afe3b20; resolved:0-->
It seems like we should acquire the sampling lock here just as we do elsewhere, since we may have a race with self.current_state_index (this may not actually happen because int32s are atomic in most architectures, but we have decided to not rely on this in other parts of this file).

pabloem replied:
<!--thread_id:cc_158386112_t; commit:3e784cbd5e110c68b56b1b4349510d712afe3b20; resolved:0-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> It seems like we should acquire the sampling lock here just as we do elsewhere, since we may have a race with self.current_state_index (this may not actually happen because int32s are atomic in most architectures, but we have decided to not rely on this in other parts of this file).

<!--section:body-->
My intention is for this method to be used just like the get_info method. Would they not work the same?

pabloem replied:
<!--thread_id:cc_158386112_t; commit:3e784cbd5e110c68b56b1b4349510d712afe3b20; resolved:0-->
<!--section:context-quote-->
> **pabloem** wrote:
> My intention is for this method to be used just like the get_info method. Would they not work the same?

<!--section:body-->
If this makes sense, I'll mark as done.

pabloem commented on sdks/python/apache_beam/runners/worker/opcounters.py:
```diff
@@ -24,11 +24,99 @@
+
+  This class is designed to track consumption of side inputs across fused steps.
+  We represent a side input as a declaring step, and an input index, which are
+  represented as an io_target.
```
<!--thread_id:cc_158386099_t; commit:1e64e1f8b880a9fe4450c84cf1c7c5694eb3567e; resolved:0-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Do we define what we mean by an `io_target` elsewhere?
> 
> Please add extra newline after paragraph.

<!--section:body-->
There's a small comment in counters.py about it. I can improve that if you think it's fitting.

pabloem replied:
<!--thread_id:cc_158386099_t; commit:1e64e1f8b880a9fe4450c84cf1c7c5694eb3567e; resolved:1-->
<!--section:context-quote-->
> **pabloem** wrote:
> There's a small comment in counters.py about it. I can improve that if you think it's fitting.

<!--section:body-->
Acknowledged.

pabloem commented:
@charlesccychen hi Charles. I've removed the context tracking part, and added only the bytes tracking. Appending the bytes tracker is behind an experiment flag, so the performance impact should be none with flag deactivated. I'll work further to do the context tracking.

pabloem commented:
Run Python PostCommit

pabloem commented:
Run Python PostCommit

pabloem commented:
Run Python PostCommit

pabloem commented:
Run Python PostSubmit

pabloem commented:
Run Python PostCommit

charlesccychen commented on sdks/python/apache_beam/runners/worker/sideinputs.py:
```diff
@@ -68,9 +73,36 @@ def __init__(self, sources,
+    # behind an experiment flag to test performance impact.
+    experiments = RuntimeValueProvider('experiments', str, '').get().split(',')
+    if 'sideinput_io_metrics' not in experiments:
+      return
```
<!--new_thread; commit:ee3c5cfd0adc6bf2848f2010c0beec4d9ace5123; resolved:0-->
Can you do this at the beginning of `_reader_thread` so that we do the check of the RuntimeValueProvider exactly once?

pabloem replied:
<!--thread_id:cc_163113306_t; commit:ee3c5cfd0adc6bf2848f2010c0beec4d9ace5123; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Can you do this at the beginning of `_reader_thread` so that we do the check of the RuntimeValueProvider exactly once?

<!--section:body-->
Done.

charlesccychen commented on sdks/python/apache_beam/runners/worker/operations.py:
```diff
@@ -306,7 +307,21 @@ def _read_side_inputs(self, tags_and_types):
         sources.append(si.source)
-      iterator_fn = sideinputs.get_iterator_fn_for_sources(sources)
+        # The tracking of time spend reading and bytes read from side inputs is
+        # behind an experiment flag to test performance impact.
```
<!--new_thread; commit:ee3c5cfd0adc6bf2848f2010c0beec4d9ace5123; resolved:0-->
"test its performance impact"

pabloem replied:
<!--thread_id:cc_163113307_t; commit:ee3c5cfd0adc6bf2848f2010c0beec4d9ace5123; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> "test its performance impact"

<!--section:body-->
Done.

charlesccychen commented on sdks/python/apache_beam/runners/worker/operations.py:
```diff
@@ -306,7 +307,21 @@ def _read_side_inputs(self, tags_and_types):
+        # behind an experiment flag to test performance impact.
+        experiments = RuntimeValueProvider(
+            'experiments', str, '').get().split(',')
+        if 'sideinput_io_metrics' in experiments:
```
<!--new_thread; commit:ee3c5cfd0adc6bf2848f2010c0beec4d9ace5123; resolved:0-->
Can you do this outside the loop so we don't need to do the RuntimeValueProvider call each time?

pabloem replied:
<!--thread_id:cc_163113309_t; commit:ee3c5cfd0adc6bf2848f2010c0beec4d9ace5123; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Can you do this outside the loop so we don't need to do the RuntimeValueProvider call each time?

<!--section:body-->
Done.

charlesccychen commented on sdks/python/apache_beam/runners/worker/opcounters.py:
```diff
@@ -24,11 +24,118 @@
+    # We check the current state to create the internal counters.
+    self.check_step()
+
+  def check_step(self):
```
<!--new_thread; commit:9c3a73fcac33b56a026e4ee575722c94c34cf64f; resolved:0-->
Do you think it could be clearer to rename this to something like `update_current_step()`?  (given how you expect this feature to progress)

pabloem replied:
<!--thread_id:cc_163113313_t; commit:9c3a73fcac33b56a026e4ee575722c94c34cf64f; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Do you think it could be clearer to rename this to something like `update_current_step()`?  (given how you expect this feature to progress)

<!--section:body-->
Sure, that sounds fine.

charlesccychen commented on sdks/python/apache_beam/runners/worker/opcounters.py:
```diff
@@ -24,11 +24,118 @@
+    self.input_index = input_index
+
+    # Side inputs are set up within the start state of the first receiving
+    # state.
```
<!--new_thread; commit:9c3a73fcac33b56a026e4ee575722c94c34cf64f; resolved:0-->
first receiving "step"?

also, remove newline, or add an extra newline here.

pabloem replied:
<!--thread_id:cc_163113315_t; commit:9c3a73fcac33b56a026e4ee575722c94c34cf64f; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> first receiving "step"?
> 
> also, remove newline, or add an extra newline here.

<!--section:body-->
Done.

charlesccychen commented on sdks/python/apache_beam/runners/worker/opcounters.py:
```diff
@@ -24,11 +24,118 @@
+
+  This class is designed to track consumption of side inputs across fused steps.
+  We represent a side input as a declaring step, and an input index, which are
+  represented as an io_target.
```
<!--new_thread; commit:9c3a73fcac33b56a026e4ee575722c94c34cf64f; resolved:0-->
Can you point to the place you say you defined "io_target" in the comment here?

pabloem replied:
<!--thread_id:cc_163113317_t; commit:9c3a73fcac33b56a026e4ee575722c94c34cf64f; resolved:0-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Can you point to the place you say you defined "io_target" in the comment here?

<!--section:body-->
It is not very well detailed, it seems. I can detail it further.
https://github.com/apache/beam/blob/master/sdks/python/apache_beam/utils/counters.py#L31

pabloem replied:
<!--thread_id:cc_163113317_t; commit:9c3a73fcac33b56a026e4ee575722c94c34cf64f; resolved:1-->
<!--section:context-quote-->
> **pabloem** wrote:
> It is not very well detailed, it seems. I can detail it further.
> https://github.com/apache/beam/blob/master/sdks/python/apache_beam/utils/counters.py#L31

<!--section:body-->
I've added documentation for io target to counters.py. Check it out.

charlesccychen commented on sdks/python/apache_beam/runners/worker/opcounters.pxd:
```diff
@@ -21,6 +21,22 @@ cimport libc.stdint
 
 
+cdef class TransformIOCounter(object):
+  cpdef add_bytes_read(self, libc.stdint.int64_t n)
```
<!--new_thread; commit:9c3a73fcac33b56a026e4ee575722c94c34cf64f; resolved:0-->
The `check_step` method isn't declared here; can you add it?

pabloem replied:
<!--thread_id:cc_163113319_t; commit:9c3a73fcac33b56a026e4ee575722c94c34cf64f; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> The `check_step` method isn't declared here; can you add it?

<!--section:body-->
Done.

charlesccychen commented on sdks/python/apache_beam/runners/worker/opcounters.py:
```diff
@@ -24,11 +24,118 @@
+    pass
+
+  def add_bytes_read(self, count):
+    pass
```
<!--new_thread; commit:9c3a73fcac33b56a026e4ee575722c94c34cf64f; resolved:0-->
Move to before `__exit__` to be consistent with `SideInputReadCounter` below.

pabloem replied:
<!--thread_id:cc_163113320_t; commit:9c3a73fcac33b56a026e4ee575722c94c34cf64f; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Move to before `__exit__` to be consistent with `SideInputReadCounter` below.

<!--section:body-->
Done.

charlesccychen commented on sdks/python/apache_beam/runners/worker/opcounters_test.py:
```diff
@@ -41,6 +46,59 @@ def __init__(self):
+    with sampler.scoped_state('step2', 'stateB'):
+      with counter:
+        counter.add_bytes_read(10)
+        sleep(0.3)
```
<!--new_thread; commit:c1c26b3fd7c3a3eba913bfa6b714db6a687754a6; resolved:0-->
Can you comment that this is just an arbitrary short interval (or otherwise, how this value was chosen)?

pabloem replied:
<!--thread_id:cc_163113326_t; commit:c1c26b3fd7c3a3eba913bfa6b714db6a687754a6; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Can you comment that this is just an arbitrary short interval (or otherwise, how this value was chosen)?

<!--section:body-->
Done.

charlesccychen commented on sdks/python/apache_beam/runners/worker/opcounters_test.py:
```diff
@@ -19,10 +19,15 @@
 import math
 import random
 import unittest
+from time import sleep
```
<!--new_thread; commit:c1c26b3fd7c3a3eba913bfa6b714db6a687754a6; resolved:0-->
Can you just import `time` and use `time.sleep()`?  This is the style in the rest of the codebase.

pabloem replied:
<!--thread_id:cc_163113327_t; commit:c1c26b3fd7c3a3eba913bfa6b714db6a687754a6; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Can you just import `time` and use `time.sleep()`?  This is the style in the rest of the codebase.

<!--section:body-->
Done.

charlesccychen commented on sdks/python/apache_beam/runners/worker/statesampler.pyx:
```diff
@@ -65,38 +71,73 @@ cdef inline int64_t get_nsec_time() nogil:
+
+  def __init__(self, state_name, transition_count, time_since_transition):
+    self.state_name = state_name
+    self.transition_count = transition_count
```
<!--thread_id:cc_158386112_t; commit:3e784cbd5e110c68b56b1b4349510d712afe3b20; resolved:1-->
<!--section:context-quote-->
> **pabloem** wrote:
> If this makes sense, I'll mark as done.

<!--section:body-->
Done.

pabloem commented:
Run Python PostCommit

charlesccychen commented on sdks/python/apache_beam/utils/counters.py:
```diff
@@ -28,18 +28,45 @@
 
 from apache_beam.transforms import cy_combiners
 
-# Information identifying the IO being measured by a counter.
```
<!--new_thread; commit:edb8389d02012005d37366b9443052d8238fdec9; resolved:0-->
Can you keep this line?  The added lines clarify the intent, but this line is still helpful as an overall description of `IOTargetName`.

pabloem replied:
<!--thread_id:cc_163387834_t; commit:edb8389d02012005d37366b9443052d8238fdec9; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Can you keep this line?  The added lines clarify the intent, but this line is still helpful as an overall description of `IOTargetName`.

<!--section:body-->
SGTM.

charlesccychen commented on sdks/python/apache_beam/runners/worker/opcounters.py:
```diff
@@ -24,11 +24,117 @@
+
+  This class is designed to track consumption of side inputs across fused steps.
+  We represent a side input as a declaring step, and an input index, which are
+  represented as an io_target.
```
<!--new_thread; commit:ac6abcff9e381c9345f7aaeabdf8f83998f6074c; resolved:0-->
I'm a little confused because this sentence seems to imply that the input index is an object of type `io_target`, which seems to clash with the documentation in `__init__` below (which is otherwise very clear).

pabloem replied:
<!--thread_id:cc_163387836_t; commit:ac6abcff9e381c9345f7aaeabdf8f83998f6074c; resolved:0-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> I'm a little confused because this sentence seems to imply that the input index is an object of type `io_target`, which seems to clash with the documentation in `__init__` below (which is otherwise very clear).

<!--section:body-->
Hmm I see the point. Yes, io_target is only used later in the CounterName. I've decided to document it there, and not mention it here.

pabloem replied:
<!--thread_id:cc_163387836_t; commit:ac6abcff9e381c9345f7aaeabdf8f83998f6074c; resolved:1-->
<!--section:context-quote-->
> **pabloem** wrote:
> Hmm I see the point. Yes, io_target is only used later in the CounterName. I've decided to document it there, and not mention it here.

<!--section:body-->
Acknowledged.

charlesccychen commented on sdks/python/apache_beam/runners/worker/opcounters.py:
```diff
@@ -24,11 +24,117 @@
+
+    The side input is uniquely identified by (declaring_step, input_index);
+    where declaring_step is the step that receives the PCollectionView as a
+    side input, and input_index is the index of the PCView within the list of
```
<!--new_thread; commit:ac6abcff9e381c9345f7aaeabdf8f83998f6074c; resolved:0-->
Please don't abbreviate `PCollectionView`.

pabloem replied:
<!--thread_id:cc_163387839_t; commit:ac6abcff9e381c9345f7aaeabdf8f83998f6074c; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Please don't abbreviate `PCollectionView`.

<!--section:body-->
Done.

charlesccychen commented on sdks/python/apache_beam/runners/worker/operations.py:
```diff
@@ -286,6 +287,10 @@ def _read_side_inputs(self, tags_and_types):
     assert self.side_input_maps is None
 
+    # Get experiments active in the worker to check for side input metrics exp.
+    experiments = RuntimeValueProvider(
```
<!--new_thread; commit:ac6abcff9e381c9345f7aaeabdf8f83998f6074c; resolved:0-->
Can you make this a `set()` to match what you have in sideinputs.py?

pabloem replied:
<!--thread_id:cc_163387840_t; commit:ac6abcff9e381c9345f7aaeabdf8f83998f6074c; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Can you make this a `set()` to match what you have in sideinputs.py?

<!--section:body-->
Done.

pabloem commented:
I can squash the commits if you think it'd be useful, Ahmet. Let me know.

aaltay commented on sdks/python/apache_beam/runners/worker/opcounters_test.py:
```diff
@@ -41,6 +47,55 @@ def __init__(self):
+    with sampler.scoped_state('step2', 'stateB'):
+      with counter:
+        counter.add_bytes_read(10)
+        time.sleep(0.1)  # An arbitrary short interval.
```
What are we testing by this sleep?

pabloem replied:
<!--thread_id:cc_163728407_t; commit:b68f9f26428de8ed2f34c0919ee91a1662402d86; resolved:1-->
<!--section:context-quote-->
> **aaltay** wrote:
> What are we testing by this sleep?

<!--section:body-->
The idea is to check that time was sampled under the time counter, but I guess this can be removed from now, while we experiment with the time counter and performance..

aaltay commented on sdks/python/apache_beam/runners/worker/operations.py:
```diff
@@ -286,6 +287,10 @@ def _read_side_inputs(self, tags_and_types):
     assert self.side_input_maps is None
 
+    # Get experiments active in the worker to check for side input metrics exp.
+    experiments = set(RuntimeValueProvider(
```
Should we add RuntimeValueProvider.get(option_name) as a class method to avoid creating dummy instance like this?

pabloem replied:
<!--thread_id:cc_163728724_t; commit:b68f9f26428de8ed2f34c0919ee91a1662402d86; resolved:0-->
<!--section:context-quote-->
> **aaltay** wrote:
> Should we add RuntimeValueProvider.get(option_name) as a class method to avoid creating dummy instance like this?

<!--section:body-->
I don't believe that instance creation represents a serious performance cost here. This is only incurred once per side input-consuming operation in a map task. I'll add it if you think it's a must.

aaltay replied:
<!--thread_id:cc_163728724_t; commit:b68f9f26428de8ed2f34c0919ee91a1662402d86; resolved:1-->
<!--section:context-quote-->
> **pabloem** wrote:
> I don't believe that instance creation represents a serious performance cost here. This is only incurred once per side input-consuming operation in a map task. I'll add it if you think it's a must.

<!--section:body-->
I was thinking more from a code style perspective of this repeated code. Maybe you can create a JIRA to improve this.

aaltay commented on sdks/python/apache_beam/runners/worker/sideinputs.py:
```diff
@@ -68,9 +73,29 @@ def __init__(self, sources,
+      if is_record_size:
+        self.read_counter.add_bytes_read(record_size)
+
+    if self.read_counter and isinstance(reader, observable.ObservableMixin):
```
Would `self.read_counter` be ever None? Is this a necessary check?

Is it an error if the reader is not a instance of ObservableMixin?

pabloem replied:
<!--thread_id:cc_163729384_t; commit:b68f9f26428de8ed2f34c0919ee91a1662402d86; resolved:1-->
<!--section:context-quote-->
> **aaltay** wrote:
> Would `self.read_counter` be ever None? Is this a necessary check?
> 
> Is it an error if the reader is not a instance of ObservableMixin?

<!--section:body-->
Currently, read_counter won't be None. I've removed the check.
And no, not all readers are subclass of ObservableMixin. Only the Avro Reader used in Dataflow - but other readers do not implement this.

pabloem commented on sdks/python/apache_beam/runners/worker/opcounters_test.py:
```diff
@@ -41,6 +47,55 @@ def __init__(self):
+
+  def setUp(self):
+    if not statesampler.FAST_SAMPLER:
+      raise SkipTest('State sampler not compiled.')
```
<!--thread_id:cc_163698968_t; commit:b68f9f26428de8ed2f34c0919ee91a1662402d86; resolved:1-->
<!--section:context-quote-->
> **aaltay** wrote:
> Have you tried using unittest.skipTest() (https://docs.python.org/2/library/unittest.html#unittest.TestCase.skipTest). We are using that in a few other places.

<!--section:body-->
With the new pure python state sampler implementation, skipping is no longer necessary.

pabloem commented on sdks/python/apache_beam/runners/worker/opcounters.py:
```diff
@@ -24,11 +24,119 @@
+    data-accessor (such as a side input / shuffle Iterable) down to the
+    next step in a stage.
+    """
+    pass
```
<!--thread_id:cc_163696527_t; commit:b68f9f26428de8ed2f34c0919ee91a1662402d86; resolved:0-->
<!--section:context-quote-->
> **aaltay** wrote:
> Should we instead raise a not implemented error for the base methods?

<!--section:body-->
I plan to use the interface as a No-op counter further down the line. I can raise the exceptions here, and make a No-op subclass if you'd rather have it that way.

aaltay replied:
<!--thread_id:cc_163696527_t; commit:b68f9f26428de8ed2f34c0919ee91a1662402d86; resolved:1-->
<!--section:context-quote-->
> **pabloem** wrote:
> I plan to use the interface as a No-op counter further down the line. I can raise the exceptions here, and make a No-op subclass if you'd rather have it that way.

<!--section:body-->
Sounds good to keep it as a noop counter. (Maybe add a comment about this in a future PR.)

pabloem commented:
Run Python PostCommit
